services/admin/src/server.test.ts

```diff
@@ -144,6 +144,71 @@ test('DELETE /admin/elections/:electionId', async () => {
   await request(app).get('/admin/elections').expect(200, []);
 });
 
+test('GET /admin/elections/:electionId/cvr-files happy path', async () => {
+  const cvrFiles: Admin.CastVoteRecordFileRecord[] = [
+    {
+      id: 'cvr-file-2',
+      createdAt: new Date().toISOString(),
+      electionId: 'test-election-2',
+      exportTimestamp: '2021-10-24T00:30:14.513Z',
+      filename: 'cvrs-2.jsonl',
+      precinctIds: ['precinct-1', 'precinct-2'],
+      scannerIds: ['scanner-1', 'scanner-2'],
+      sha256Hash: 'file-2-hash',
+    },
+    {
+      id: 'cvr-file-1',
+      createdAt: new Date().toISOString(),
+      electionId: 'test-election-2',
+      exportTimestamp: '2021-09-02T22:27:58.327Z',
+      filename: 'cvrs-1.jsonl',
+      precinctIds: ['precinct-2'],
+      scannerIds: ['scanner-1'],
+      sha256Hash: 'file-1-hash',
+    },
+  ];
+
+  jest
+    .spyOn(workspace.store, 'getCvrFiles')
+    .mockImplementationOnce((electionId: Id) => {
+      expect(electionId).toEqual('test-election-2');
+
+      return cvrFiles;
+    });
```

Comment:
yeah, i was waffling on this one a bit - curious what @jonahkagan's thoughts are on this too, since we've talked recently about test patterns.

you bring up a good point about "unit" tests - i was approaching these in the context of wanting to write self-contained, hermetic unit tests, but admittedly the code under test here isn't particularly written with that sort of testing in mind.

i think we'd ideally make the dependencies of the api handler more easily injectable, so that the test has the ability to modify dependency behaviour and verify that the code reacts appropriately

```ts
const requestHandler = newHandler(mockStore, mockExternalService);
const app = createApp(requestHandler, ...);
const response = await request(app).get('...');
expect(response).etc()
```

but, in the absence of something along those lines, filling in with a jest mock instead.
trying to avoid repeating the exact same test for both the store method and the handler and also avoiding having this test breaking for something as simple as the store changing how results are sorted (for example)

that all being said, i'd rather stay consistent if this strays too far from what we think is ideal as a team

Reply:
Reflecting on my recent leanings around tests, I've been trying to test against the interfaces that I'm most confident aren't going to be refactored often and then mock as little as possible. That way as things behind the scenes change, the tests are still useful.

E.g. in Arlo basically all the server tests run a db and hit the API with nothing mocked. There are a few unit tests for particularly complex code (e.g. audit math). I'm free to change whatever I want inside the server.

So in this example, I wouldn't write tests for the store at all.

Reply:
I'm liking this as a guiding principle for the services! As long as they stay simple enough (handler -> business logic/storage) and don't have too many layers in between that make it difficult to get the system into a specific state for a test case.

And can't think of any obvious pitfalls in terms of "can I write a bug that gets caught with one approach, but not the other".

Will spend some time looking at the store tests to see if moving them to endpoint tests is straightforward 🚀

Reply:
I'm liking this as a guiding principle for the services! As long as they stay simple enough (handler -> business logic/storage) and don't have too many layers in between that make it difficult to get the system into a specific state for a test case.

And can't think of any obvious pitfalls in terms of "can I write a bug that gets caught with one approach, but not the other".

Will spend some time looking at the store tests to see if moving them to endpoint tests is straightforward 🚀

Reply:
Awesome! I'm even envisioning that we will be able to write those tests with the autogenerated client library so they don't have to deal with the transport layer directly